From 75deabb33ea48553eb12b75d65739292c5202209 Mon Sep 17 00:00:00 2001
From: Carl Smedstad <carl.smedstad@protonmail.com>
Date: Fri, 15 Nov 2024 21:14:09 +0100
Subject: [PATCH] Replace dependency zc.lockfile with filelock

The filelock library fills the same role but is more popular &
established.
See: https://github.com/tox-dev/filelock
---
 cherrypy/lib/sessions.py | 14 ++++++++------
 setup.py                 |  2 +-
 2 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/cherrypy/lib/sessions.py b/cherrypy/lib/sessions.py
index f0d2678b4..ad29c49b2 100644
--- a/cherrypy/lib/sessions.py
+++ b/cherrypy/lib/sessions.py
@@ -108,7 +108,7 @@
 import binascii
 import pickle
 
-import zc.lockfile
+from filelock import FileLock, Timeout
 
 import cherrypy
 from cherrypy.lib import httputil
@@ -579,19 +579,21 @@ def acquire_lock(self, path=None):
         path += self.LOCK_SUFFIX
         checker = locking.LockChecker(self.id, self.lock_timeout)
         while not checker.expired():
+            lock = FileLock(path)
+            timeout = 0.1
             try:
-                self.lock = zc.lockfile.LockFile(path)
-            except zc.lockfile.LockError:
-                time.sleep(0.1)
-            else:
+                lock.acquire(timeout=timeout)
+                self.lock = lock
                 break
+            except Timeout:
+                time.sleep(timeout)
         self.locked = True
         if self.debug:
             cherrypy.log('Lock acquired.', 'TOOLS.SESSIONS')
 
     def release_lock(self, path=None):
         """Release the lock on the currently-loaded session data."""
-        self.lock.close()
+        self.lock.release()
         self.locked = False
 
     def clean_up(self):
diff --git a/setup.py b/setup.py
index 53f3bc645..031f47f9f 100644
--- a/setup.py
+++ b/setup.py
@@ -68,7 +68,7 @@
         'cheroot>=8.2.1',
         'portend>=2.1.1',
         'more_itertools',
-        'zc.lockfile',
+        'filelock',
         'jaraco.collections',
         'importlib-metadata; python_version<="3.7"',
     ],
